controller/src/main/java/org/jboss/as/controller/ListAttributeDefinition.java

```diff
@@ -61,6 +63,14 @@ public ParameterValidator getElementValidator() {
         return elementValidator;
     }
 
+    /**
+     * Returns an AttributeDefinition describing the content of the list.
+     * @return an AttributeDefinition describing the content of the list.
+     */
+    public AttributeDefinition getValueAttributeDefinition() {
+        throw ROOT_LOGGER.missingListAttributeValueType(getName());
```

Comment:
The Javadoc must mention when this exception can be raised.

When did you catch it actually? How can the code using a ListAttributeDefinition assert if calling this method is valid or not?

I'm concerned by having this method raises an exception by default. If the fundamental definition of a list can not provide an attributedefinition for its item, it does not belong to this class.

It seems to me that your change would need an intermediate class between `ListAttributeDefinition` and subclasses that *can* provide such an attribute definition (which is the case for all the subclasses defined in wildfly-core, I've not checked in wildfly full).

Having an intermediate `AttributeDefinition-based ListAttributeDefintion` would also simplify a lot the various subclasses we have (Object, Primitive, String, Simple)

Reply:
The exception is not caught. Trying to configure a list that is not properly defined using YAML will fail.

Note that this only concerns ListAttributeDefinitions that are configurable. So you mean that we have a list which values aren't definable by AttributeDefinition but that we want to be able to configure using YAML ? In such a case it will fail, exception or not, so I think that throwing a standard exception is clearer.
Just FYI we have several such ListAttributeDefinitions in full for runtime attributes.

Reply:
What make a `list` configurable with YAML?
The presence of a valueType (directly for primitive or though an attributedefinition for object list)?
Is is valid to have a list without a valueType?

Reply:
If it is not valid to have a list w/o a valid "value definition", I think the thing to do is have this method return null, and then YamlConfigurationExtension processListAttribute can fail if null is returned. IOW move the exception throwing to the code that can't deal with a situation, and don't have it in the ListAD code which doesn't care whether there is a value definition.

From the calling code POV that allows callers who don't know what ListAD subclass they are dealing with to do flow control using a null check instead of a try/catch. If the only place that would ever call this was YamlConfigurationExtension that distinction wouldn't matter much (as it doesn't want to catch) but methods in core API classes like ListAD should be robust when other uses we haven't thought of appear.

Reply:
I hate returning 'null' tbh as this leads to NPE (even if i agree that a I could use a checked exception in that case.